src/content/docs/en/recipes/i18n.mdx

```diff
@@ -0,0 +1,322 @@
+        <p>by {page.data.author} at {formattedDate}</p>
+        <Content/>
+        ```
+      </Fragment>
+
+      <Fragment slot="ssr">
+        In [SSR mode](/en/guides/server-side-rendering/), fetch the requested entry directly:
+
+        ```astro
+        //src/pages/[lang]/blog/[...slug].astro
+        ---
+        import { getEntryBySlug } from 'astro:content';
+
+        const { lang, slug } = Astro.params;
+        const page = await getEntryBySlug('blog', `${lang}/${slug}`);
+
+        if (!page) {
+          return Astro.redirect('/404');
+        }
+
+        const formattedDate = page.data.date.toLocaleString(lang);
+        const { Content, headings } = await page.render();
+        ---
+        <h1>{page.data.title}</h1>
+        <p>by {page.data.author} at {formattedDate}</p>
+        <Content/>
+        ```
+      </Fragment>
+    </StaticSsrTabs>
+
+    📚 Read more about [dynamic routing](/en/core-concepts/routing/#dynamic-routes).
+
+### Translate UI strings
+
+Create dictionaries of terms to translate the labels for UI elements around your site. This allows your visitors to experience your site fully in their language.
+
+1. Create a `src/i18n/ui.ts` file to store your translation strings in:
+
+    ```ts
+    // src/i18n/ui.ts
+    export const languages = {
+      en: 'English',
+      fr: 'français',
+    };
+
+    export const defaultLang = 'en';
+    
+    export const ui = {
+      en: {
+        'nav.home': 'Home',
+        'nav.about': 'About',
+        'nav.twitter': 'Twitter',
+      },
+      fr: {
+        'nav.home': 'Acceuil',
+        'nav.about': 'À propos',
+      },
+    } as const;
+    ```
+    
+2. Create a helper function to detect the page language based on the current URL and a helper to let us get translations strings for different parts of our UI in `src/i18n/utils.ts`:
+
+    ```js
+    // src/i18n/utils.ts
+    import { ui, defaultLang } from './ui';
+    
+    export function getLangFromUrl(url: URL) {
+      const [, lang] = Astro.url.pathname.split('/');
+      return lang || defaultLang;
+    }
+    
+    export function useTranslations(lang: keyof typeof ui) {
+      return function t(key: keyof typeof ui[defaultLang]) {
+        return ui[lang][key] || ui[defaultLang][key];
+      }
+    }
+    ```
+
+    :::note[Did you notice?]
+    In step 1, the `nav.twitter` string was not translated to French. The `useTranslations` helper here will return the default language’s value if a key is not translated so that French users will still see “Twitter”.
+    :::
+    
+3. Use our helpers to choose the UI string that corresponds to the current language. For example, a nav component might look like:
+
+    ```astro 
+    ---
+    // src/components/Nav.astro
+    import { getLangFromUrl, useTranslations } from '../i18n/utils';
+    
+    const lang = getLangFromUrl(Astro.url);
+    const t = useTranslations(lang);
+    ---
+    <ul>
+        <li>
+            <a href={`/${lang}/home/`}>
```

Comment:
astro-i18n uses a function to generate routes which I like, similar DX as strings so it makes more sense in my mind than getting into string interpolation https://github.com/alexandre-fernandez/astro-i18n#translated-routing-function

Reply:
Nice! Yeah, we decided to keep this recipe real simple and avoid a specific library but we’ve linked astro-i18n at the end of the page for people to explore.